clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -2075,19 +2075,47 @@ protected DelegateConsumerRecordResult delegateConsumerRecord(
                   kafkaClusterId,
                   beforeProcessingRecordTimestampNs);
             } else {
-              if (controlMessageType == START_OF_SEGMENT
-                  && Arrays.equals(consumerRecord.getKey().getKey(), KafkaKey.HEART_BEAT.getKey())) {
-                LeaderProducerCallback callback = createProducerCallback(
-                    consumerRecord,
-                    partitionConsumptionState,
-                    leaderProducedRecordContext,
-                    subPartition,
-                    kafkaUrl,
-                    beforeProcessingRecordTimestampNs);
-                LeaderMetadataWrapper leaderMetadataWrapper =
-                    new LeaderMetadataWrapper(consumerRecord.getOffset(), kafkaClusterId);
-                PubSubTopicPartition topicPartition = new PubSubTopicPartitionImpl(getVersionTopic(), subPartition);
-                veniceWriter.get().sendHeartbeat(topicPartition, callback, leaderMetadataWrapper);
+              if (controlMessageType == START_OF_SEGMENT) {
+                if (Arrays.equals(consumerRecord.getKey().getKey(), KafkaKey.HEART_BEAT.getKey())) {
+                  LeaderProducerCallback callback = createProducerCallback(
+                      consumerRecord,
+                      partitionConsumptionState,
+                      leaderProducedRecordContext,
+                      subPartition,
+                      kafkaUrl,
+                      beforeProcessingRecordTimestampNs);
+                  LeaderMetadataWrapper leaderMetadataWrapper =
+                      new LeaderMetadataWrapper(consumerRecord.getOffset(), kafkaClusterId);
+                  PubSubTopicPartition topicPartition = new PubSubTopicPartitionImpl(getVersionTopic(), subPartition);
+                  // Leaders forward HB SOS message to local VT with updated isLeaderCompleted header
+                  veniceWriter.get()
+                      .sendHeartbeat(
+                          topicPartition,
+                          callback,
+                          leaderMetadataWrapper,
+                          true,
+                          consumerRecord.getValue().producerMetadata.messageTimestamp,
+                          partitionConsumptionState.isCompletionReported());
+                } else {
+                  // Leaders forward SOS message to local VT: add isLeaderCompleted Header
+                  produceToLocalKafka(
```

Comment:
Do we really want to forward SOS messages? I thought we would only tag SOS messages from the leaders writer itself.

Reply:
Makes sense to me and made the changes. We can continue to iterate on this in this PR or part 2 PR based on the teams opinion as well.